helix-view/src/editor.rs

```diff
@@ -1715,3 +1715,106 @@ fn try_restore_indent(doc: &mut Document, view: &mut View) {
         doc.apply(&transaction, view.id);
     }
 }
+
+static EXPAND_ARGS_REGEXP: once_cell::sync::Lazy<helix_core::regex::Regex> =
+    once_cell::sync::Lazy::new(|| {
+        helix_core::regex::Regex::new(r"%(\w+)\{([^{}]*(\{[^{}]*\}[^{}]*)*)\}").unwrap()
+    });
+
+pub fn expand_args<'a>(editor: &Editor, args: &'a str) -> anyhow::Result<Cow<'a, str>> {
+    let (view, doc) = current_ref!(editor);
+    let shell = &editor.config().shell;
+
+    replace_all(
+        once_cell::sync::Lazy::force(&EXPAND_ARGS_REGEXP),
+        Cow::Borrowed(args),
+        move |keyword, body| match keyword.trim() {
+            "val" => match body.trim() {
+                "filename" => Ok(doc
+                    .path()
+                    .and_then(|it| it.to_str())
+                    .unwrap_or(crate::document::SCRATCH_BUFFER_NAME)
+                    .to_string()),
+                "dirname" => doc
+                    .path()
+                    .and_then(|p| p.parent())
+                    .and_then(std::path::Path::to_str)
+                    .map_or(
+                        Err(anyhow::anyhow!("Current buffer has no path or parent")),
+                        |v| Ok(v.to_string()),
+                    ),
+                "line_number" => Ok((doc
+                    .selection(view.id)
+                    .primary()
+                    .cursor_line(doc.text().slice(..))
+                    + 1)
+                .to_string()),
+                _ => anyhow::bail!("Unknown variable: {body}"),
+            },
+            "sh" => {
+                let result = tokio::task::block_in_place(move || {
+                    helix_lsp::block_on(async move {
+                        let args = &expand_args(editor, body)?[..];
+
+                        let mut command = tokio::process::Command::new(&shell[0]);
+                        command.args(&shell[1..]).arg(args);
+
+                        let output = command
+                            .output()
+                            .await
+                            .map_err(|_| anyhow::anyhow!("Shell command failed: {args}"))?;
+
+                        if output.status.success() {
+                            String::from_utf8(output.stdout)
+                                .map_err(|_| anyhow::anyhow!("Process did not output valid UTF-8"))
+                        } else if output.stderr.is_empty() {
+                            Err(anyhow::anyhow!("Shell command failed: {args}"))
+                        } else {
+                            let stderr = String::from_utf8_lossy(&output.stderr);
+
+                            Err(anyhow::anyhow!("{stderr}"))
+                        }
+                    })
+                });
+
+                result.map_err(|it| {
+                    log::error!("{}", it.to_string());
+
+                    it
+                })
```

Comment:
I don't think we need to log here since we're returning and handling the Err variant. `Editor::set_error` already logs each error message